### DISCUSSION TITLE
Public Tickets

### DISCUSSION THREAD

**Discussion by jorge-leon** [django-helpdesk/django-helpdesk#1191] [General]
Public pages allow not logged in clients and not-staff users to submit and view tickets.  This discussion collects related issues and observations.

Feedback is appreciated.

### Staff Users

Staff users have two ways to see their own tickets:
- Dashboard
- My Tickets

The last one is superfluos
 
Proposed solution: Suppress the "My Tickets" menu entry for staff users.

### Missing user menu

When using "My Tickets" or a public ticket view while logged in, the user menu is not shown, e.g. log out is not possible.

Proposed solution: Add respective code.

### Security of public Ticket URLs

Public Ticket URL's are GET (obviously) and expose the secret key via the URL. Note also  attachments are guarded by the secret key.

Proposed solution:
- Create a public ticket identifier e.g. by using the secret key as salt for encrypting the ticket number. Use this public identifier in the URL
- Store files under encrypted filenames which use the secret key as salt.

### Attachments on public ticket submission page

Be able to selectively switch off attachments for public ticket submission.
Some site might tickets from anyone on the Internet, attachments from registered users, but not attachments from anyone on the Internet.

### Show selected closed/resolved tickets in Knowledge Base

Some ticket systems allow knowledge bases to build upon user tickets: A resolved ticket could be marked "public" and then be shown in the KB-Item where it was created.  Other users could then see the resolution provided by
staff.

### Missing information on public ticket view

Following is not shown on public ticket view:
- Attachment list
- Dependencies, Parents
- TicketCC (Copies to)
- Checklists

### Sidenotes

The following is not directly related but popped up while reviewing public views. Should be moved out eventuelly to other discussions or issue reports.

- "All Tickets" should have a Filter on "Submitter E-Mail", where one could filter for "myself".
- Reports on Submitter Email would be interesting
- When clicking on the KB-Item for a ticket (also) link to the KB-Item, now only to the search.
- Reopened tickets show still the "Resolution" field in the description table.
- New ticket submission requires the description, although the model does not require the field on purpose.
- When creating a ticket from a KB-Item, the Queue is not set, although a KB-Category has a (default) queue.

**Comment by uhurusurfa** [comment]
In general I think what you have written above makes complete sense and I would support PR's that add the above functionality.

I think the "My Tickets" view for staff makes sense if Heldesk is used in an internal environment where they not only process tickets from cutomers/other staff but also submit their own tickets for help they need and therefore we just need to make the staff view show tickets they are assigned top process and the my tickets view only tickets they have submitted for processing.

**Comment by samsplunks** [comment]
Totally agree with the ticket security part.
There is also a form where you can POST the ticket secret key but have not used it yet.

For ticket attachments (and security), I have overridden a django view for authenticated user access to serve /helpdesk/attachments files.

I also wanted to introduce a setting to set the default "My Tickets" filter view.

**Comment by ronflima** [comment]
I found that a non-staff user cannot update the ticket with comments. The view that updates the ticket explicitly check if the user is staff. It is possible to let the user update tickets by setting `HELPDESK_ALLOW_NON_STAFF_TICKET_UPDATE ` parameter. However, this parameter is too broad and will allow a user to interact with other support tickets. In a commercial software, this is not desired at all. 

So, there should be a way to users to update tickets they have submitted, but in a limited way: add more attachments and add comments. Both are currently not supported although very easy to customize at the installation (by customizing templates and the view itself). 

Do you think it is worth adding a new setting for this kind of scenario? If so, I'll work in a PR for such a thing.

**Reply by uhurusurfa** [reply]
Please feel free to raise a PR to support the enhancement.

**Reply by ronflima** [reply]
I'm starting to study the current solution so I can propose assertive changes. It looks like we have a gap in terms of concepts: what are supposed to be public tickets and how public tickets could be managed by those who opened it. In my opinion, there are the following actors playing a role here:

- The **reporter**: The person who created the ticket at a first place. It can be an external person (public tickets), an internal person (someone that has a system account) or a staff member (scenario where a staff member opens a ticket to another staff member to service it).
- The **ticket owner**: The person who will service the ticket. 

Therefore, the ticket life cycle is took care by both parties. The reporter can close the ticket at any time, as soon as it consider the problem solved or found a workaround. The ticket owner can transit the ticket through other states during the analysis/resolution workflow. But also can close the ticket. The reporter should not be able to change a ticket to state different from `closed`. However, it should see state changes in order to follow up ticket resolution.

**Comment by uhurusurfa** [comment]
The above is correct as a basic premise.

There are a number of odd implementations in this app that will need to be addressed and possibly there will be some breaking changes that will ultimately make this app easier to enhance and upgrade especially around permissions.

A key problem I see is that the app was designed as a plug in to other projects and as such does not attempt to provide a User model.
As a consequence, the app relies on the implementer to provide a User model but we already allocate the "is_staff" field provided on the Django default User model for determining who are staff users and non-staff users. If the app is plugged into a project that has already utilised the "is_staff" for some other purpose then it is not going to work from a security perspective.

Because there is only an email field that tracks the reporter/submitter of the ticket (Ticket.submitter_email), we use the email field on the User model to try and link the reporter to an actual user but if that does not exist then all we know about the submitter is an email address.
There are a number of issues posted in this p[roject that are focussed on improved security of user access to tickets and related information and if this app is plugged into an app that already has some form of security implementation for user access to ticket  creation, update and delete then clearly it would be best to use that projects access control implementation.

It would be better if we created a role based system for access control to define user privileges that are independent of the User model and design it in such a way as to make it possible to adapt the Helpdesk user privileges into other implementations or just use the default implementation we design within the app. This should be reltively simple if we provide an interface for priveleges and then an implementation module as default allowing other projects to implement a wrapper using the interface we define to map the actual implementation to their own already available implementation.

**Reply by ronflima** [reply]
I was thinking about that. What about using the Django model-based system for permissions? So, we can rely on the user model the same way Django relies on (through `get_user_model()`) but using the user's `has_perm()` method. It is already there, it is really well-tested and stable. Unless you think that we should be completely independent from `django.contrib.auth` package. 

It feels to me that the django permissions system is pretty straightforward and can provide the level of granularity we need for letting users edit, create, or even comment tickets.

**Reply by uhurusurfa** [reply]
Currently the views are not model specific so whilst it is feasible to use Django permissions, it gets quite messy becasue many of the views update and create records in multiple models. This makes it difficult for end users to maintain permissions without having detailed knowledge of the models that each view affects. It may be more appropriate to use [Django Guardian](https://django-guardian.readthedocs.io/) and align a view with an object level permission.
I have never actually used the Django permissions system and not particularly enthused by it as it has .signficant limitations in my view but perhaps if we define clear permissions requirements and how they should work at a functional level then we can map that into a technical implementation using a mix of Django permissions and Django Guardian

**Reply by ronflima** [reply]
Yeah, Django permissions can be complicated to use without proper modeling. You have a very good point: the user of this app can be confused. I'll look at Django Guardian.
